Lock missing dep fixes

iarna opened the pull request:
This has a handful of fixes:

1. It introduces a new `package-lock.json` field, called `requires`, which tracks which modules a given module requires.
2. It fixes #16839 which was caused by not having this information available, particularly when git dependencies were involved.
3. It fixes #16866, allowing the `package.json` to trump the `package-lock.json`.
4. It stops calling `fetch-package-metadata.addBundled` on fake children (this was harmless, because bundleDependencies would never be set, but also pointless.)
5. `npm ls` now loads the shrinkwrap, which opens the door to showing a full tree of dependencies even when nothing is yet installed. (It doesn't do that yet though.)

iarna committed:
refresh-package-json: Children aren't fake after reading real pjson data

Fixes: #16839

iarna committed:
install: Don't prune if any children are fake

Also prune before we save, as we skipped the earlier pruning and we don't
want to save a bogus `package-lock`.

Fixes: #16839

iarna committed:
shrinkwrap: Save requires information to shrinkwrap/lock

Fixes: #16839

iarna committed:
install/deps: Cleanup doesChildVersionMatch implementation

This removes a no longer appropriate check of `fromShrinkwrap` and
a duplicated code path when using `_from`.

iarna committed:
inflate-shrinkwrap: Stop calling addBundle on locked deps

iarna committed:
inflate-shrinkwrap: Inflate dependencies from shrinkwrap

Fixes: #16839

iarna committed:
inflate-shrinkwrap: Let the installer fix invalid lockfiles

Fixes: #16866

iarna committed:
spec/package-lock: packageIntegrity is no longer a thing

iarna committed:
spec/package-lock: Document requires

iarna committed:
install/deps: Fakechildren w/o requires always count as semver matches

Fixes: #16839

iarna committed:
install/deps: If we can't find a dependecies match, try the lock's requires

Fixes: #16839

iarna committed:
ls: Take shrinkwrap into account in ls

iarna committed:
package-lock: Update npm's own package lock

zkat commented on lib/install/inflate-shrinkwrap.js:
```diff
@@ -26,7 +26,8 @@ module.exports = function (tree, swdeps, opts, finishInflating) {
   }
-  tree.loaded = true
+  tree.loaded = false
+  tree.hasRequiresFromLock = Object.keys(swdeps).every((d) => swdeps[d].requires)
```
shouldn't this be `.some()`? You only need to know if there was _one_, right?

iarna replied:
well... no, if any of them are missing then we can't assume we know anything about the relationships between the modules.

iarna replied:
At one point during the process of writing this it was possible to produce an install that only had partial resolved information. Buuut…

That doesn't seem to be the case any more. So this can become `some` and we can stop including `requires` on every node.

Relatedly, I'm finding that `npm install xyz` or `npm rm xyz` when you have a legacy package-lock but nothing installed fails to properly upgrade the package-lock, eg:

```
$ node ~/code/npm rm null

added 190 packages and removed 1 package in 6.671s
```

zkat commented on package-lock.json:
```diff
@@ -6,113 +6,163 @@
       "resolved": "https://registry.npmjs.org/abbrev/-/abbrev-1.1.0.tgz",
-      "integrity": "sha1-0FVMIlZjbi9W58LlrRg/hZQo2B8="
+      "integrity": "sha1-0FVMIlZjbi9W58LlrRg/hZQo2B8=",
+      "requires": {}
```
wait, was your solution to unconditionally add a `requires:{}`? Mehhhh. I don't like this kinda noop crud :\

iarna replied:
Well, I'm open to alternatives.

zkat commented on lib/shrinkwrap.js:
```diff
@@ -133,13 +127,28 @@ function shrinkwrapDeps (deps, top, tree, seen) {
     }
     if (childIsOnlyDev) pkginfo.dev = true
     if (isOptional(child)) pkginfo.optional = true
+    pkginfo.requires = {}
```
meh @ always having `requires`

zkat commented on lib/install/mutate-into-logical-tree.js:
```diff
@@ -81,7 +81,12 @@ function translateTree_ (tree, seen) {
+    if (child.fakeChild) {
+      dep.missing = true
+      dep.optional = child.package._optional
+      dep.requiredBy = child.package._spec
```
I don't get this line...

iarna replied:
This is where we're producing `read-installed`'s output for `ls` to consume. In it's format `requiredBy` is the spec that required the missing dep.

zkat commented on lib/install/inflate-shrinkwrap.js:
```diff
@@ -105,43 +105,40 @@ function makeFakeChild (name, topPath, tree, sw, requested) {
     _where: topPath,
-    _args: [[requested.toString(), topPath]]
+    _args: [[requested.toString(), topPath]],
+    dependencies: sw.requires
```
this feels so mischievous 😈

zkat reviewed (commented):
Added some comments. Nothing really bad jumps out at me. I'll take a closer look once I've gotten some sleep and/or can actually talk through this and test it out.

iarna committed:
shrinkwrap: Only include requires when something was required

iarna committed:
install: Insist on full tree when mutating with a package.lock

That is, if you have a package-lock but haven't installed anything yet then
we're going to insist that all of the bits be there. This is necessary in order
to be able to cleanly upgrade old package-lock files. This means if you type
`npm install foo` or `npm rm foo` and don't have a `node_modules` but do have
a `package-lock.json` it's gonna install everything from the
`package-lock.json` in addition to making your change.

The one thing we won't do is update your package-lock from your
package.json.  We only do that when you request a full install.

This does not change the behavior if you don't have a lock file.

iarna committed:
install: Guard against missing ideal tree

iarna committed:
install: Include flag at top if the lock contains requires

iarna committed:
tests: Make more lenient

zkat commented:
@iarna when I use `npm ls`, I notice that we get all those nice unmet dependencies, which is awesome. There's two things, tho:

1. All unmet dependencies are reported as required by the _root_ package, not by their actual logical parent.
2. I feel like it's redundant for us to spit out individual errors/warnings when we're already printing the tree that shows exactly which ones are missing/invalid/etc. We could probably just collapse that all into a _single_ warning/error that says "some dependencies are missing or invalid" before doing exit 1.

iarna commented:
@zkat If it's showing you anything beyond the first level that's a bug. (It doesn't do that for me in my examples…)

We can bikeshed the details of how this looks different showing the full tree later, imo.

legodude17 commented on doc/spec/package-lock.md:
```diff
@@ -117,9 +112,17 @@ on the current platform.
 
-The dependencies of this dependency, exactly as at the top level.
+Exactly like `dependencies` at the top level, this is a list of modules to
+isntall in the `node_modules` of this module.
```
Is that supposed to say `isntall`?

iarna replied:
nope, ty

legodude17 commented on doc/spec/package-lock.md:
```diff
@@ -28,15 +28,6 @@ The version of the package this is a package-lock for. This must match what's in
 An integer version, starting at `1` with the version number of this document
 whose semantics were used when generating this `package-lock.json`.
 
-### packageIntegrity *(new)*
```
Why was this removed? I am confused.

iarna replied:
This was a hash of the `package.json` which would result in churn and merge conflicts on the `package-lock.json`. It was never implemented and the consensus of the discussion is that it was a bad idea.

legodude17 commented on test/tap/shrinkwrap-local-dependency.js:
```diff
@@ -85,7 +85,7 @@ test('shrinkwrap uses resolved with file: on local deps', function (t) {
       t.equal(code, 0, 'npm exited normally')
       var data = fs.readFileSync(path.join(testdir, 'npm-shrinkwrap.json'), { encoding: 'utf8' })
-      t.deepEqual(
+      t.like(
```
Is this an update to `tap` or a design choice?

iarna replied:
`deepEqual` requires exact matching of the datastructures. `like` let's you compare an incomplete datastructure against a complete one. Where they share keys they have to match.

So this change was made to allow changes to the `package-lock` format that aren't related to this test to happen w/o breaking the test.

legodude17 reviewed (commented):
:package:

iarna committed:
test: get rid of pointless ifError calls in ls-production-and-dev

iarna committed:
test: Update fixture lockfile formats

iarna committed:
shrinkwrap: Move is-only-optional and is-only-dev out to their own files

iarna committed:
install/node: Never fill in fromBundle from _inBundle

We were overriding this, so it wasn't an issue, but it can't be right as
_inBundle is a boolean and fromBundle is a package object.

iarna committed:
extract: Read bundles if any children are fromBundle

Previously we checked pkg.package.bundleDependencies which isn't available
to fake children.

iarna committed:
deps: Stop avoiding walking shrinkwrapped dev deps

We walk all deps now.

iarna committed:
inflate-shrinkwrap: Stop filtering deps here

That is stop filtering based on `--prod`, `--only=dev`, etc
In future this will happen in `diff-trees`

iarna committed:
inflate-shrinkwrap: Guard against empty/missing dependencies

iarna committed:
inflate-shrinkwrap: Make fake children for bundled deps too

iarna committed:
inflate-shrinkwrap: Standardize how and which _ props we add to package metadata

iarna committed:
diff-trees: Filter out prod and dev deps here when requested

Previously this happened in various places, `inflate-shrinkwrap` and in
`install` in particular.  As of this writing it's still happening in
`install`—but we'll change that soon.

iarna committed:
install: Only add removed children to summary if there's not already an action for them

zkat reviewed (approved):
LGTM! 💯 🐑 🚀 💥 🔥 😱

iarna committed:
save: Don't update the package.json when modifying packages that don't go there

Fixes: #17141

ntwb commented:
Oh well, this didn't work, once Travis CI tests are run a second time with caching enabled npm continues to crash 😢 

• https://travis-ci.org/WordPress-Coding-Standards/eslint-plugin-wordpress/jobs/249240419

Reference: https://github.com/npm/npm/issues/17419
